### PR TITLE
Fix: Closing bottom sheet and removing FAB cause assertion failure

### LINKED ISSUE #128562: Closing bottom sheet and removing FAB cause assertion failure
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Display bottom sheet using `showBottomSheet` and `FloatingActionButton` in stateful scaffold
2. `PersistentBottomSheetController.close()` and `floatingActionButton: null` at the same time

### Expected results

Assertion failure should not happen.

### Actual results

```
════════ Exception caught by animation library ═════════════════════════════════
The following assertion was thrown while notifying status listeners for AnimationController:
'package:flutter/src/material/scaffold.dart': Failed assertion: line 1362 pos 16: 'widget.currentController.status == AnimationStatus.dismissed': is not true.
```

This assertion is not needed when removing FAV (`widget.child == null`).

### Code sample

<details><summary>Code sample</summary>

Remove bottom sheet and FAV at the same time with `count == 3`.

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MaterialApp(home: _CounterScaffold()));

class _CounterScaffold extends StatefulWidget {
  const _CounterScaffold();

  @override
  State<_CounterScaffold> createState() => _CounterScaffoldState();
}

class _CounterScaffoldState extends State<_CounterScaffold> {
  int _count = 0;

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: const Text('Sample Code')),
        body: _CounterBody(count: _count),
        floatingActionButton: _count < 3
            ? FloatingActionButton(
                onPressed: () => setState(() => _count++),
                tooltip: 'Increment Counter',
                child: const Icon(Icons.add),
              )
            : null,
      );
}

class _CounterBody extends StatefulWidget {
  final int count;

  const _CounterBody({required this.count});

  @override
  State<_CounterBody> createState() => _CounterBodyState();
}

class _CounterBodyState extends State<_CounterBody> {
  PersistentBottomSheetController? _controller;

  @override
  void didUpdateWidget(_CounterBody oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (oldWidget.count == 0 && widget.count > 0) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        _controller = showBottomSheet(
          context: context,
          builder: (_) => Container(height: 200, color: Colors.amber),
        );
      });
    }

    if (_controller != null && widget.count >= 3) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        _controller!.close();
        _controller = null;
      });
    }
  }

  @override
  Widget build(BuildContext context) =>
      Center(child: Text('You have pressed the button ${widget.count} times.'));
}
```

</details>

### Screenshots or Video

_No response_

### Logs

<details><summary>Logs</summary>

```console
════════ Exception caught by animation library ═════════════════════════════════
The following assertion was thrown while notifying status listeners for AnimationController:
'package:flutter/src/material/scaffold.dart': Failed assertion: line 1362 pos 16: 'widget.currentController.status == AnimationStatus.dismissed': is not true.

Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md

When the exception was thrown, this was the stack
#2      _FloatingActionButtonTransitionState._handlePreviousAnimationStatusChanged.<anonymous closure>
#3      State.setState
#4      _FloatingActionButtonTransitionState._handlePreviousAnimationStatusChanged
#5      AnimationLocalStatusListenersMixin.notifyStatusListeners
#6      AnimationController._checkStatusChanged
#7      AnimationController._tick
#8      Ticker._tick
#9      SchedulerBinding._invokeFrameCallback
#10     SchedulerBinding.handleBeginFrame.<anonymous closure>
#11     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#12     SchedulerBinding.handleBeginFrame
#13     SchedulerBinding._handleBeginFrame
#14     _invoke1 (dart:ui/hooks.dart:158:13)
#15     PlatformDispatcher._beginFrame (dart:ui/platform_dispatcher.dart:337:5)
#16     _beginFrame (dart:ui/hooks.dart:101:31)
(elided 2 frames from class _AssertionError)
The AnimationController notifying status listeners was: AnimationController#24868(⏮ 0.000; paused)
════════════════════════════════════════════════════════════════════════════════
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.4, on macOS 13.3.1 22E261 darwin-arm64, locale ja-JP)
    • Flutter version 3.10.4 on channel stable at /opt/homebrew/Caskroom/flutter/3.10.0/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 682aa387cf (3 days ago), 2023-06-05 18:04:56 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/masatoshitsushima/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode-14.3.0.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.78.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.66.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.106

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by utisam**
Fixes #128562

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by TahaTesser** [comment]
Thanks for fixing this! @utisam 

We'll need a test for this fix.

**Comment by utisam** [comment]
@TahaTesser Test was added at 02404bc

The assertion failure was caused on the latest master channel.

File: packages/flutter/test/material/scaffold_test.dart [hunk 1]
```diff
@@ -2800,6 +2800,43 @@ void main() {
...
+      );
+    }));
+
+    setState(() => show = false);
+
+    await tester.pump(); // start animation
+    await tester.pump(const Duration(seconds: 1));
+
+    expect(controller?.closed, completes);
+    expect(find.byType(FloatingActionButton), findsNothing);
```
Comment (TahaTesser) [inline packages/flutter/test/material/scaffold_test.dart:2838]: I'm not really sure if this is the correct way to test this.

When I comment these lines. I get errors

```dart
    // setState(() => show = false);

    // await tester.pump(); // start animation
    // await tester.pump(const Duration(seconds: 1));

    // expect(controller?.closed, completes);
    // expect(find.byType(FloatingActionButton), findsNothing);
```

File: packages/flutter/test/material/scaffold_test.dart [hunk 2]
```diff
@@ -2800,6 +2800,43 @@ void main() {
...
+    PersistentBottomSheetController<void>? controller;
+    bool show = true;
+    late StateSetter setState;
+
+    await tester.pumpWidget(StatefulBuilder(builder: (_, StateSetter stateSetter) {
+      setState = stateSetter;
+      return MaterialApp(
+        home: Scaffold(
+          body: Builder(builder: (BuildContext context) {
+            WidgetsBinding.instance.addPostFrameCallback((_) {
```
Comment (TahaTesser) [inline packages/flutter/test/material/scaffold_test.dart:2814]: Is `addPostFrameCallback` necessary? 
It'd be preferable to add and remove the bottom sheet by pressing buttons.

**Comment by TahaTesser** [comment]
@Piinks 
This seems to be a risky change and test. Do you know anyone who has expertise in this area to review this PR?

**Review by Piinks — COMMENTED** [review]
Thanks for raising this @TahaTesser. Welcome @utisam! Thanks for contributing! 
Can you take a look at the test suggestion below? It may be that we just need a bit more context to understand the design of the test - thank you for including one by the way. :)

File: packages/flutter/test/material/scaffold_test.dart (hunk 2 above)
Comment (utisam) [reply packages/flutter/test/material/scaffold_test.dart]: Thanks for the comments. I rewrite the test using buttons.

> Yeah, maybe can the FAB.onPressed trigger this change?

I also tried to trigger 2 changes at FAB.onPressed at once, but it does not reproduce the assertion error.

While rewriting the test, I found the problem is only that while the hiding FAB animation is running, another event starts the closing bottom animation. The reversed order (close bottom sheet → hide FAB) or both at once does not.

File: packages/flutter/test/material/scaffold_test.dart [hunk 3]
```diff
@@ -2800,6 +2800,56 @@ void main() {
...
+    expect(find.byKey(bottomSheetKey), findsOneWidget);
+
+    // Close bottom sheet while hiding FAB
+    await tester.tap(find.byType(FloatingActionButton));
+    await tester.pump(); // start animation
+    await tester.tap(find.byType(ElevatedButton));
+    await tester.pumpAndSettle(const Duration(seconds: 1));
+    expect(find.byType(FloatingActionButton), findsNothing);
+    expect(find.byKey(bottomSheetKey), findsNothing);
+  });
```
Comment (TahaTesser) [inline packages/flutter/test/material/scaffold_test.dart:2852]: This looks great! Just some minor improvements are needed.

This updates the title, fixes the code format, most importantly adds an assertion check.

```suggestion
  testWidgets(
    "Closing bottom sheet & removing FAB at the same time doesn't throw assertion",
    (WidgetTester tester) async {
      final Key bottomSheetKey = UniqueKey();
      PersistentBottomSheetController<void>? controller;
      bool show = true;

      await tester.pumpWidget(StatefulBuilder(
        builder: (_, StateSetter setState) => MaterialApp(
          home: Scaffold(
            body: Center(
            child: Builder(
                builder: (BuildContext context) => ElevatedButton(
                  onPressed: () {
                    if (controller == null) {
                      controller = showBottomSheet(
                        context: context,
                        builder: (_) => Container(
                        key: bottomSheetKey,
                        height: 200,
                      ));
                    } else {
                      controller!.close();
                      controller = null;
                    }
                  },
                  child: const Text('BottomSheet'),
                )),
            ),
            floatingActionButton: show
            ? FloatingActionButton(
                onPressed: () => setState(() => show = false),
              )
            : null,
          ),
        ),
      ));

      // Show bottom sheet.
      await tester.tap(find.byType(ElevatedButton));
      await tester.pumpAndSettle(const Duration(seconds: 1));

      // Bottom sheet and FAB are visible.
      expect(find.byType(FloatingActionButton), findsOneWidget);
      expect(find.byKey(bottomSheetKey), findsOneWidget);

      // Close bottom sheet while removing FAB.
      await tester.tap(find.byType(FloatingActionButton));
      await tester.pump(); // start animation
      await tester.tap(find.byType(ElevatedButton));
      // Let the animation finish.
      await tester.pumpAndSettle(const Duration(seconds: 1));

      // Bottom sheet and FAB are gone.
      expect(find.byType(FloatingActionButton), findsNothing);
      expect(find.byKey(bottomSheetKey), findsNothing);

      // No exception is thrown.
      expect(tester.takeException(), isNull);
  });

```

**Review by TahaTesser — CHANGES_REQUESTED** [review]

**Review by TahaTesser — APPROVED** [review]
LGTM

**Comment by TahaTesser** [comment]
We'll need a second review from @Piinks before landing this.

**Review by Piinks — APPROVED** [review]
Just a couple of style guide nits from me, which I was able to just tweak here on github. Thanks again for contributing @utisam! LGTM!

**Pull request merged**